metricflow/model/validations/identifiers.py

```diff
@@ -119,3 +123,86 @@ def validate_model(model: UserConfiguredModel) -> List[ValidationIssue]:  # noqa
+                SubIdentifierContext(
+                    data_source_name=data_source.name,
+                    identifier_reference=identifier.name,
+                    sub_identifier_names=tuple(IdentifierConsistencyRule._get_sub_identifier_names(identifier)),
+                )
+            )
+        return contexts
+
+    @staticmethod
+    @validate_safely(whats_being_done="running model validation to ensure identifiers have consistent sub-identifiers")
+    def validate_model(model: UserConfiguredModel) -> List[ValidationIssueType]:  # noqa: D
+        issues: List[ValidationIssueType] = []
+        # The expected sub identifiers that an identifier has.
+        identifier_to_sub_identifiers: OrderedDict[str, Tuple[str, ...]] = OrderedDict()
+        # The identifiers that have different sub-identifiers in different data sources.
+        identifiers_with_inconsistent_sub_identifiers = []
+
+        identifier_contexts = []
+        issues = []
+        for data_source in model.data_sources or []:
+            sub_identifier_contexts = IdentifierConsistencyRule._get_sub_identifier_context(data_source)
+            for sub_identifier_context in sub_identifier_contexts:
+                identifier_contexts.append(sub_identifier_context)
+
+                identifier_reference = sub_identifier_context.identifier_reference
+                if identifier_reference.element_name not in identifier_to_sub_identifiers:
+                    identifier_to_sub_identifiers[
+                        identifier_reference.element_name
+                    ] = sub_identifier_context.sub_identifier_names
+
+                elif (
+                    identifier_to_sub_identifiers[identifier_reference.element_name]
+                    != sub_identifier_context.sub_identifier_names
+                ):
+                    identifiers_with_inconsistent_sub_identifiers.append(identifier_reference)
+
+        for identifier_reference in identifiers_with_inconsistent_sub_identifiers:
+            for identifier_context in identifier_contexts:
+                if identifier_context.identifier_reference.element_name == identifier_reference.element_name:
+                    issues.append(
+                        ValidationWarning(
+                            model_object_reference=ValidationIssue.make_object_reference(
+                                data_source_name=identifier_context.data_source_name,
+                                identifier_name=identifier_context.identifier_reference.element_name,
+                            ),
+                            message=(
+                                f"Identifier '{identifier_reference.element_name}' does not have consistent sub-identifiers "
+                                f"throughout the model. In data source '{identifier_context.data_source_name}', "
+                                f"it has sub-identifiers {list(identifier_context.sub_identifier_names)}."
+                            ),
```

Comment:
It'd be nice to provide the sub-identifiers of the original in the error message as well, but it's a bit tricky since there could be more than one mismatch. Right now the user has to go look at the data source definition to find the original set of sub-identifiers, right? 

I'd rather have the original set of sub-identifiers repeated in every error then not have them at all, so if that information isn't included in the set of warnings anywhere let's add them somehow.

Reply:
New commit includes all the definitions of the offending identifiers, including in which data source they are defined.